File: examples/FunctionDefaults.py
#!/usr/bin/env python3

# pylint: disable = bad-whitespace
# pylint: disable = dangerous-default-value
# pylint: disable = eval-used
# pylint: disable = invalid-name
# pylint: disable = missing-docstring

# -------------------
# FunctionDefaults.py
# -------------------

def f (x, y, z=5) :
    return [x, y, z]

def test1 () :
    assert f(2, 3)    == [2, 3, 5]
    assert f(2, 3, 4) == [2, 3, 4]

# def g (x, y=5, z) : # SyntaxError: non-default argument follows default argument
#     return [x, y, z]

def g (x=5, y=6, z=7) :
    return [x, y, z]

def test2 () :
    assert g()        == [5, 6, 7]
    assert g(2)       == [2, 6, 7]
    assert g(2, 3)    == [2, 3, 7]
    assert g(2, 3, 4) == [2, 3, 4]
    assert g(2, z=4)  == [2, 6, 4]

def h1 (x=[]) : # mutable default
    x += [2]
    return x

def test3 () :
    assert h1()    == [2]
    assert h1()    == [2, 2]
    assert h1([1]) == [1, 2]
    assert h1()    == [2, 2, 2]
    assert h1([1]) == [1, 2]

def h2 (x=()) : # immutable default
    x += (2,)
    return x

def test4 () :
    assert h2()     == (2,)
    assert h2()     == (2,)
    assert h2((1,)) == (1, 2)
    assert h2()     == (2,)
    assert h2((1,)) == (1, 2)

def h3 (x=None) :
    if x is None :
        x = []
    x += [2]
    return x

def test5 () :
    assert h3()     == [2]
    assert h3()     == [2]
    assert h3([1])  == [1, 2]
    assert h3()     == [2]
    assert h3([1])  == [1, 2]
    assert h3(None) == [2]

def main () :
    print("FunctionDefaults.py")
    for i in range(5) :
        eval("test" + str(i + 1) + "()")
    print("Done.")

if __name__ == "__main__" : # pragma: no cover
    main()

File: examples/FunctionDefaults.js
#!/usr/bin/env node

/* jshint
    esversion: 6,
    evil:      true,
    -W138:     true  // Regular parameters should not come after default parameters.
*/

// -------------------
// FunctionDefaults.js
// -------------------

// https://www.w3schools.com/js/js_function_parameters.asp

"use strict";

const assert = require('assert');
const _      = require('lodash');

function f (x, y, z) {
    return [x, y, z];}

function test1 () {
    assert(_.isEqual(f(),        [undefined, undefined, undefined]));
    assert(_.isEqual(f(2),       [2,         undefined, undefined]));
    assert(_.isEqual(f(2, 3),    [2,         3,         undefined]));
    assert(_.isEqual(f(2, 3, 4), [2,         3,         4]));}

function g1 (x, y, z=5) {
    return [x, y, z];}

function test2 () {
    assert(_.isEqual(g1(),        [undefined, undefined, 5]));
    assert(_.isEqual(g1(2),       [2,         undefined, 5]));
    assert(_.isEqual(g1(2, 3),    [2,         3,         5]));
    assert(_.isEqual(g1(2, 3, 4), [2,         3,         4]));}

function g2 (x, y=5, z) {
    return [x, y, z];}

function test3 () {
    assert(_.isEqual(g2(),        [undefined, 5, undefined]));
    assert(_.isEqual(g2(2),       [2,         5, undefined]));
    assert(_.isEqual(g2(2, 3),    [2,         3, undefined]));
    assert(_.isEqual(g2(2, 3, 4), [2,         3, 4]));}

function h (x=[]) {
    x.push(2);
    return x;}

function test4 () {
    assert(_.isEqual(h(), [2]));
    assert(_.isEqual(h(), [2]));}

function main () {
    console.log("FunctionDefaults.js");
    for (let i of _.range(4))
        eval("test" + (i + 1) + "()");
    console.log("Done.");}

main();
